File: app/services/validation.py
from __future__ import annotations

import re

from app.models.enums import ContactMethod, ServiceCategory


class ValidationError(ValueError):
    pass


NAME_MIN = 2
NAME_MAX = 80
DESCRIPTION_MIN = 10
DESCRIPTION_MAX = 1000
PHONE_MIN_DIGITS = 8
PHONE_MAX_DIGITS = 15


def validate_name(raw: str) -> str:
    name = " ".join((raw or "").split())
    if not NAME_MIN <= len(name) <= NAME_MAX:
        raise ValidationError("Enter a full name using 2–80 characters.")
    if re.search(r"https?://|www\.", name, flags=re.IGNORECASE):
        raise ValidationError("Enter a name, not a link.")
    if not any(ch.isalpha() for ch in name):
        raise ValidationError("Enter a full name using letters.")
    return name


def normalize_phone(raw: str) -> str:
    text = (raw or "").strip()
    if text.startswith("00"):
        text = f"+{text[2:]}"
    digits = re.sub(r"\D", "", text)
    if not PHONE_MIN_DIGITS <= len(digits) <= PHONE_MAX_DIGITS:
        raise ValidationError("Enter a phone number with 8 to 15 digits.")
    return f"+{digits}"


def validate_description(raw: str) -> str:
    text = (raw or "").strip()
    if len(text) < DESCRIPTION_MIN:
        raise ValidationError("Describe the project in at least 10 characters.")
    if len(text) > DESCRIPTION_MAX:
        raise ValidationError("Keep the description under 1,000 characters.")
    return text


def parse_category(value: str) -> ServiceCategory:
    try:
        return ServiceCategory(value)
    except ValueError as exc:
        raise ValidationError("Choose a service from the list.") from exc


def parse_contact_method(value: str) -> ContactMethod:
    try:
        return ContactMethod(value)
    except ValueError as exc:
        raise ValidationError("Choose a contact method from the list.") from exc

File: showcase/src/lib/leads/validation.ts
const NAME_RE = /^[\p{L}][\p{L}\s.'-]{1,79}$/u;

export function normalizePhone(raw: string): string {
  const trimmed = raw.trim();
  let value = trimmed.replace(/[^\d+]/g, "");
  if (value.startsWith("00")) {
    value = `+${value.slice(2)}`;
  }
  const digits = value.replace(/\D/g, "");
  if (digits.length < 8 || digits.length > 15) {
    throw new Error("Enter a phone number with 8 to 15 digits.");
  }
  return `+${digits}`;
}

export function validateName(raw: string): string {
  const name = raw.trim().replace(/\s+/g, " ");
  if (!NAME_RE.test(name)) {
    throw new Error("Enter a full name using 2–80 letters.");
  }
  return name;
}

export function validateDescription(raw: string): string {
  const text = raw.trim();
  if (text.length < 10) {
    throw new Error("Describe the project in at least 10 characters.");
  }
  if (text.length > 1000) {
    throw new Error("Keep the description under 1,000 characters.");
  }
  return text;
}

export function generatePublicId(date = new Date()): string {
  const stamp = date.toISOString().slice(2, 10).replaceAll("-", "");
  const bytes = new Uint8Array(2);
  crypto.getRandomValues(bytes);
  const suffix = Array.from(bytes, (b) => b.toString(16).padStart(2, "0"))
    .join("")
    .toUpperCase();
  return `LF-${stamp}-${suffix}`;
}
